source/common/common/mutex_tracer.cc

```diff
@@ -0,0 +1,28 @@
+#include "common/common/mutex_tracer.h"
+
+#include <memory>
+
+namespace Envoy {
+
+MutexTracer* MutexTracer::GetTracer() {
+  static MutexTracer global_tracer;
+  return &global_tracer;
+}
+
+void MutexTracer::ContentionHook(const char* msg, const void* obj, int64_t wait_cycles) {
+  GetTracer()->RecordContention(msg, obj, wait_cycles);
+}
+
+void MutexTracer::Reset() {
+  num_contentions_.store(0, order_);
+  current_wait_cycles_.store(0, order_);
+  lifetime_wait_cycles_.store(0, order_);
+}
+
+void MutexTracer::RecordContention(const char*, const void*, int64_t wait_cycles) {
+  num_contentions_.fetch_add(1, order_);
+  current_wait_cycles_.store(wait_cycles, order_);
```

Comment:
I'm kind of interested in what this would be used for: just the most recent number of wait cycles. I think 5-second/1-minute/1-hour running average might be more useful, but I don't think we have that infra in our stats system.

Reply:
Agreed, that would be extremely cool. Maybe I'll take a TODO() on it.